src/input_stream.rs

```diff
@@ -0,0 +1,180 @@
+// Copyright 2013-2017, The Gtk-rs Project Developers.
+// See the COPYRIGHT file at the top-level directory of this distribution.
+// Licensed under the MIT license, see the LICENSE file or <http://opensource.org/licenses/MIT>
+
+
+use Cancellable;
+use Error;
+use ffi;
+#[cfg(any(feature = "v2_34", feature = "dox"))]
+use glib;
+use glib::object::IsA;
+use glib::translate::*;
+use glib_ffi;
+use gobject_ffi;
+use std::mem;
+use std::ptr;
+use InputStream;
+
+pub trait InputStreamExtManual {
+    fn read<'a, B: AsMut<[u8]>, P: Into<Option<&'a Cancellable>>>(&self, buffer: B, cancellable: P) -> Result<isize, Error>;
+
+    fn read_all<'a, P: Into<Option<&'a Cancellable>>>(&self, buffer: Vec<u8>, cancellable: P) -> Result<Vec<u8>, (Error, Vec<u8>)>;
+
+    #[cfg(any(feature = "v2_44", feature = "dox"))]
+    fn read_all_async<'a, P: Into<Option<&'a Cancellable>>, Q: Fn(Vec<u8>, Option<Error>) + Send + 'static>(&self, buffer: Vec<u8>, io_priority: i32, cancellable: P, callback: Q);
```

Comment:
@antoyo might have an opinion on this one for the futures-glib stuff. I didn't look at that, but I assume you require the closures to get a `Result` passed, which is then converted into `Future::Item` and `Future::Error`? That might be an argument for making this a `Result<Vec<u8>, (Vec<u8>, Error)>` instead.

And generally, having a common "shape" for the async functions could be beneficial in the future for some generic higher-level thing around them.

Reply:
Minor question here what better `(Error, Vec<u8>)` or `(Vec<u8>, Error)` ?

Reply:
Hm... or maybe even `Result<(Vec<u8>, Option<Error>), Error>`? The error case with read bytes is not *really* an error

Reply:
Then when we return `Err(err)` part?

Reply:
When 0 bytes were read and there was an error

Reply:
IMHO we need return buffer in all error cases  to reuse be possible.
So `Result<Vec<u8>, (Vec<u8>, Error)>` or with reversed error tuple is only options.

Reply:
`Result<(Vec<u8>, Option<Error>), (Vec<u8>, Error)>`?

Reply:
Maybe. So `read_all` need same change